add base protocol upgrade spec

jeremyletang opened the pull request:
Base specs for protocol upgrades without LNL restarts.

jeremyletang committed:
add base protocol upgrade spec

jeremyletang committed:
add acceptance criteria

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+- then decide of an actual date which would give a week for the validator to asynchronously update their node.
+
+This upgrade went quite smoothly without any incident, although it was possible as the fix was located in a single place, and wasn't impacting much of the state
+(meaning no changes in the protobug were required for example.)
```
```suggestion
(meaning no changes in the protobuf files were required for example.)
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+
+The solution employed at the time was to:
+- implement a hotfix
+- keep the hotfix behing a guard until a given time, once the time is reachd network parameters would be dispatched.
```
```suggestion
- keep the hotfix behind a guard until a given time, once the time is reached network parameters would be dispatched.
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+### Prior example of updgrading the nodes asynchronously
+
+Back in December 2020, vega proceeded to a LNL restore. Unfortunately, a bug in the code prevented the dispatch of the network parameters after the restore. This left
+the network in a semi invalid state where the network parameters where defined by a the ones from the genesis block instead of the one from the checkpoint from the previous network.
```
```suggestion
the network in a semi invalid state where the network parameters were defined by the ones from the genesis block instead of the one from the checkpoint from the previous network.
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+
+### How different protocol proceed
+
+Other protocol e.g Ethereum proceed to updates in an asynchronous manner. Usually a new version of the protocol is made available, an hardcoded block height is set
```
```suggestion
Other protocol e.g Ethereum carry out updates in an asynchronous manner. Usually a new version of the protocol is made available, a hardcoded block height is set
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+### How different protocol proceed
+
+Other protocol e.g Ethereum proceed to updates in an asynchronous manner. Usually a new version of the protocol is made available, an hardcoded block height is set
+at which a new code path will be enabled. If enough nodes runner have deployed the updated code, then the network will continue with the new code path. Others nodes
```
```suggestion
at which a new code path will be enabled. If enough node runners have deployed the updated code, then the network will continue with the new code path. Others nodes
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+## Asynchronous upgrades mecanism
+
+One of the main challenge with upgrading the network both asynchronously and without a restart / downtime is
+the management of the state, and possible uncompatible changes between the state of two version of the protocol.
```
```suggestion
the management of the state, and possible incompatible changes between the state of two version of the protocol.
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+One of the main challenge with upgrading the network both asynchronously and without a restart / downtime is
+the management of the state, and possible uncompatible changes between the state of two version of the protocol.
+
+In order to prevent any state incompatiblity introduce a new framework to propose upgrades of the protocol, this will allow the node to be stopped,
```
```suggestion
In order to prevent any state incompatibility, this specification introduces a new framework to propose upgrades of the protocol, which will allow the node to be stopped,
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+the management of the state, and possible uncompatible changes between the state of two version of the protocol.
+
+In order to prevent any state incompatiblity introduce a new framework to propose upgrades of the protocol, this will allow the node to be stopped,
+and restart from the block it stopped using the snapshots or to replay the whole chain. When being restarted
```
```suggestion
and restart from the block it stopped at using the snapshots or by replaying the whole chain. When being restarted
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+In order to prevent any state incompatiblity introduce a new framework to propose upgrades of the protocol, this will allow the node to be stopped,
+and restart from the block it stopped using the snapshots or to replay the whole chain. When being restarted
+the new node will contain both the previous version of the protocol in use, but also a new version that the
+network should start using at a given block. Upong restart, the node will try to send a transaction to
```
```suggestion
network should start using at a given block. Upon restart, the node will try to send a transaction to
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+to proceed with the update.
+
+
+The upgrade requires to be done in an atomic way. To do so, the core node will need to have been compiler with
```
```suggestion
The upgrade **must** be done in an atomic way. To do so, the core node will need to have been compiled with
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+The upgrade requires to be done in an atomic way. To do so, the core node will need to have been compiler with
+the current and next version of the protocol. Once the block height for the upgrade is reached the currently
+running version of the protocol will stop executing any transaction (e.g: at the ABCI Commit call), then the
+whole state will be serialized, all the engine shut down, and the new version of the protocol will be
```
```suggestion
whole state will be serialised, all of the engines will shut down, and the new version of the protocol will be
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+the current and next version of the protocol. Once the block height for the upgrade is reached the currently
+running version of the protocol will stop executing any transaction (e.g: at the ABCI Commit call), then the
+whole state will be serialized, all the engine shut down, and the new version of the protocol will be
+instantiated, and will have to load the state previously serialized by the old version. The new version of
```
```suggestion
instantiated, and will have to load the state previously serialised by the old version. The new version of
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+running version of the protocol will stop executing any transaction (e.g: at the ABCI Commit call), then the
+whole state will be serialized, all the engine shut down, and the new version of the protocol will be
+instantiated, and will have to load the state previously serialized by the old version. The new version of
+the protocol will be in charged to adapt / convert the previous state in order to make it compatible with
```
```suggestion
the protocol will be then to convert the previous state in order to make it compatible with
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+the protocol will be in charged to adapt / convert the previous state in order to make it compatible with
+any changes in the protocol.
+
+The most main difficulty here will be the implementation of the serialisation of the whole state of the node.
```
```suggestion
The main difficulty here will be the implementation of the serialisation of the whole state of the node.
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+any changes in the protocol.
+
+The most main difficulty here will be the implementation of the serialisation of the whole state of the node.
+Fortunately we already have such feature which should work out of the box with minimal changes: the snapshots.
```
```suggestion
Fortunately we already have such feature which should work out of the box with minimal changes: [snapshots](../non-protocol-specs/0009-NP-SNAP-snapshots.md).
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+Fortunately we already have such feature which should work out of the box with minimal changes: the snapshots.
+
+
+## Versionning of the protocol
```
```suggestion
## Versioning of the protocol
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+
+### Patch upgrade
+
+This upgrade should represent very small change in the protocol, eventually a bug fix, or QOL improvement.
```
```suggestion
This upgrade should represent very small change in the protocol, for example a bug fix or a quality-of-life improvement.
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+### Patch upgrade
+
+This upgrade should represent very small change in the protocol, eventually a bug fix, or QOL improvement.
+These upgrade require no involvement from any actors of the network apart from the validator to update their node
```
```suggestion
These upgrade require no involvement from any actors of the network apart from the validators updating their nodes
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+
+### Minor upgrade
+
+These are upgrade which brings more features to the protocol, eventual fixes as well, and changes in the state
```
```suggestion
These are upgrade which bring more features to the protocol, or larger fixes, and changes in the state
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+### Minor upgrade
+
+These are upgrade which brings more features to the protocol, eventual fixes as well, and changes in the state
+produce by the protocol. They may require more works from the validators and will not be applied automatically.
```
```suggestion
produced by the protocol. They may require more work from the validators and will not be applied automatically.
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+### Major upgrade
+
+These are important changes in the protocol, possibly breaking the previously known behaviours of the protocol
+and bringing also new important feature. Due to the nature of the upgrade, a governance vote will be required
```
```suggestion
but also implementing new important features. Due to the nature of the upgrade, a governance vote will be required
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+and bringing also new important feature. Due to the nature of the upgrade, a governance vote will be required
+so token holders show of their support for the upgrade.
+
+> note: the major upgrade will need refinment, and are not planned at first.
```
```suggestion
> note: the definition of a major upgrade will need refinement, and the current version of the specification does not cover it.
```

edd commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,160 @@
+
+## Acceptance criteria
+- [ ] A new patch upgrade is made available
+  - [ ] Enough validator have upgraded their node
```
```suggestion
  - [ ] Enough validators have upgraded their node
```

edd replied:
Also this one could be clearer what 'enough' means (e.g. 2/3+1 have signalled with a X event)

edd reviewed (commented):
Good spec so far! I've gone through and fixed some basic typos, then I'll go back and add some more feedback

edd committed:
Typo/grammar fixes pass

No major changes to anything in the specification

edd committed:
Update 0070-PLUP_protocol_upgrades.md

Clarify intro

edd committed:
Update 0070-PLUP_protocol_upgrades.md

Add AC codes, add note about upgrades from versions

edd commented:
@jeremyletang looks good to me. I've fixed typos and grammatical mistakes, tidied up and coded the acceptance criteria, and added a section that basically says 'If the network rejects an upgrade, subsequent upgrades need to take that in to account'. That last one might be up for debate, but I think this is ready for review by @vegaprotocol/research

edd commented:
@barnabee In this spec, @jeremyletang added the requirement that the block height is built in to the code, rather than specified at a point where nodes upgrade their software. It feels brittle to me, and specific to a single network rather than letting us do the same upgrade across devnet/testnet/networks running on ethereum with the same upgrade, but at different points. Could you talk about why you want it compiled in?

wwestgarth commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,187 @@
+  - [ ] Enough validators have upgraded their node
+  - [ ] At least `validators.vote.required` message have been sent by the validators
+  - [ ] The block time for the upgrade is reached
+  - [ ] The network changes to the new version of the protocol (can be checked with APIs)
```
Are we going to entertain the idea that if a patch does not contain changes the break consensus, then a set of validators could be running with the same minor version, but will different patch version? Or are we just enforcing that all validators need to be running identical software, even down to the patch number?

jeremyletang replied:
Well, I'm not sure about this. I'd imagine that we would want everyone to update to the last version just in case.
Do you think it would be too much of a hassle? I imagine that a patch could actually be something that fix a potential consensus issue in a future block too.

wwestgarth replied:
I only asks because its something tendermint supports in that is separates overall version number from the "protocol version". The easiest and safest thing certainly would be to have everyone running identical code.

wwestgarth commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,187 @@
+The upgrade **must** be done in an atomic way. To do so, the core node will need to have been compiled with
+the current and next version of the protocol. Once the block height for the upgrade is reached the currently
+running version of the protocol will stop executing any transaction (e.g: at the ABCI Commit call), then the
+whole state will be serialised, all of the engines will shut down, and the new version of the protocol will be
```
This is just an observation really. If there is a bug in the code that serialising the whole state of vega (snapshots), for example a vital peice of state is not being captured, then we're screwed right? Because we won't be able to migrate that value into the new vega version.

Unless we replay the whole chain with a new version that has *just* the snapshot fix in. Maybe.

jeremyletang replied:
yes that's correct, we would miss the data, an replaying the wholke chain would be needed, and applying the fix only when we reach the point in the chain where the upgrade is supposed to be applied, then the state would be save from this point.

Vegaklaus commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,187 @@
+Other protocol e.g Ethereum carry out updates in an asynchronous manner. Usually a new version of the protocol is made available, a hardcoded block height is set
+at which a new code path will be enabled. If enough node runners have deployed the updated code, then the network will continue with the new code path. Others nodes
+which haven't updated will then fork the blockchain.
+
```
Eth may not be the best example, as it has a different behaviour (mainly, if their upgrade causes disagreement., they forc; if ours causes disagreement, the chain stops until sufficient weight supports either version. Maybe refer to how cosmos does it (https://docs.cosmos.network/master/core/upgrade.html) ?

jeremyletang replied:
Yes, I was using them as an example of keeping the chain running while processing upgrade, not specifically how they do them. The cosmos way is how we will do our upgrade indeed.

Vegaklaus commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,187 @@
+
+
+## Versioning of the protocol
+
```
Since we're (at least in theory) not the only provider of code, the version should probably also contain a hash to make sure it's uniquely defined. Otherwise, it would be possible that validators think they agree on the version (as the numbers match), but run code from different providers. Another way would be that Vega still signs the code, and thus gives their version a unique vendor ID (I'm not sure how far we do that now).
The latter might be a good idea anyhow, as that would create a well defined authentication of update code, and we can implement 4-eye security and such.

jeremyletang replied:
Well I think that ideally we don't want to vendor lock vega as the only provider of the code. We version the protocol here, and fixes to the protocol, so I would expect that version N of the protocol should behave the same accros different implementation. The vega node itself though, is version separately.

gordsport replied:
Need a way to inform the validators that they cannot upgrade from version 1 to 3 and they need to update to version 2 first then 3. This also needs to account for less simple cases where version numbers are not sequential (i.e. semver + serial number).

Vegaklaus commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,187 @@
+
+
+## Versioning of the protocol
+
```
Another point that should be clarified with versioning is the downgrade option - is it ok to downgrade if a new version has a problem, or should someone in that case provide a new version number with the older code (i.e., the code would refuse to update to a lower version number) ? I suppose given that an update code has both versions of the protocol, a downgrade would require a new code blob from us ?
(It might be good if the validators could switch back if needed; however, this would mean that there old version can parse the serialisation of the new one, which would need a extra layer of code and not be overly feasible.

jeremyletang replied:
Yea, we could decide to do that. That would be a massive difficulty, is something was to change so much that you cannot use it in a previous version. I would instead prefer to have to upgrade all the time, and for us to provide a fix version that the validators would use to upgrade to.

Vegaklaus replied:
Agree here; the only issue is that if we need a fast rollback due to an issue, we still have to be provide the functionality for the rollback version (now with the higher version number) to read the serialisation of the current version, so a fast rollback isn't always possible.

jeremyletang replied:
Well, I'd imagine a rollback would be us providing a new version, where the upgrade is the previous version and the legacy version is the upgrade, with a new layer in between to convert back the data to the previous state?

gordsport replied:
@gordsport - add risk to risk log of complex rollback if indeed possible.

@jeremyletang Can we have, in the future, an activate/deactivate process to mitigate this risk? (new spec on feature toggling?)

Vegaklaus commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,187 @@
+defined for the upgrade is reached and if enough nodes have send the transaction notifying of their intent of
+migrating to the new version of the protocol, then the upgrade is applied. If not enough transaction have been
+sent by the validators, the protocol will keep running the current version.
+
```
Those transactions are sent through the chain as well to make sure we have consensus ? And, what is enough ? 2/3 of the validator weight (which allows to run, but no longer for failures) or more ?

jeremyletang replied:
Yea that should be 2/3 of the validators weight I suppose.

Vegaklaus commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,187 @@
+- Meanwhile v`0.3.0` is released
+
+The network is now in a situation where `v0.3.0` expects the network to be upgrading from v`0.2.0`, while it would actually be upgrading from v`0.1.0`. This is expected, and upgrades must take this in to account. 
+
```
I'm not sure that's compatible with what's described in line 53; given the code has both the old and the new version, and the code needs to be able to parse and transfer the state, jumping an update should not be possible (unless I misunderstood something). 
Maybe it would be an idea to also add a serial number to upgrades or add to every upgrade which version it expects to upgrade from(so they have to come in order), and it is our job to keep track of what the network is doing and add the right serial number/code

jeremyletang replied:
well, a code release will be compatible only for upgrade from a version to another. Now if a version is not accepted in time by validators, then it will be for us to provide code again, with the same legacy version than in production, and change the upgrade path from 1 -> 2 to 1 -> 3 (meaning the convertion code has to be re-written etc, version 2 is discarded completely).

Vegaklaus commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,187 @@
+ > Note: These acceptance criteria are *incomplete*. There will likely be a governance vote from users, as well as the validator vote. For now these Acceptance Criteria test a major version in the same was a minor version upgrade.
+- [ ] A new major upgrade is made available, validators vote to update, and the update is successfully deployed (<a name="0070-PLUP-006" href="#0070-PLUP-006">0070-PLUP-006</a>)
+  - [ ] Enough validators have upgraded their node
+  - [ ] A governance proposal have been created to propose the upgrade
```
Would the governance proposal be hard-coded ? I.e., if all validators upgrade anyhow, would the code stop them ?
If it is, can the validators use the upgrade on testnet anyhow ? Do they have a way to simulate a governance proposal there easily that's compatible with the enforcement?

jeremyletang replied:
I'm not sure we are speaking of a governance proposal here. It's more of a framework internal to the protocol, just for the validators to accept or not the release manually

gordsport replied:
Do not expect this will be used right now - still needs to be specified

Vegaklaus commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,187 @@
+  - [ ] Enough validators have upgraded their node
+  - [ ] A governance proposal have been created to propose the upgrade
+  - [ ] The proposal have been rejected
+  - [ ] The block time for the upgrade is reached
```
We used block height before. Both works, but needs to be clear which one we use.

jeremyletang replied:
I don't mind as well, what's the most practical / speaking better to users?

gordsport replied:
Should we use both i.e. block height for patch and block time for major?

Do we also need an option in case we never reach the height or time due to the chain having stopped?

gordsport commented:
Requirement to add to the spec: 
https://vegaprotocol.slack.com/archives/CAHA5EX0F/p1657541187157559?thread_ts=1656933690.212839&cid=CAHA5EX0F

To help non-validators auto-update (if enabled)

fkondej commented on protocol/0070-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,187 @@
+
+### Prior example of updgrading the nodes asynchronously
+
+Back in December 2020, vega proceeded to a LNL restore. Unfortunately, a bug in the code prevented the dispatch of the network parameters after the restore. This left
```
I think it was 2021, not 2020.

ze97286 committed:
feat: initial version of the new protocol update process

gordsport committed:
chore: update filename and AC codes

Updating the file name and AC codes and the number 0070 was taken already now 0075.

Checked all files in repo to ensure no links to this spec too

gordsport committed:
fix: fix typo

gordsport committed:
fix: typo in ACs

jgsbennett commented:
###  Network Parameters and other upgrade-time-config

config.toml - manually changed only, not automated - not supplied with binary

network params - stored within binary - no use case at present

When the network is restarted with the new version, what configuration might we want to change _at exactly that time_? And what mechanisms are there to do it? @karlem mentioned there would be vega parameters , but also there might be new or existing network parameters that validators would wish to set or change at the same block the new version takes effect. Currently, Network Param proposals take effect at a given _time_, not a given block height, so I believe this spec does not allow for setting new values for new versions, and enforces us to hard-code validator chosen defaults, validators to rebuild themselves, or else for the validators to begin the network with values they are not able to control, until such a time as they can arrange for proposals to succeed.
If we choose to add extra functionality for this, we'll need more ACs to cover cases around this.



### Extra ACs / Test cases / Questions based on existing spec:

VISOR
- Vega visor automatically downloads the next version for install
- Vega visor uses the correct version, which has been pre-downloaded into the directory

Negative cases
- Propose an upgrade block in the past - should be rejected / failed check - sign via wallet / check wallet
- Propose a negative upgrade block
- Propose a 0 upgrade block
- Propose (string/other upgrade block)
- Do not supply an upgrade block in the proposal (if at all possible)

Positive
- Propose an upgrade block which ought to be the end of an epoch - 
- Propose an upgrade block which should result in a new network in the same epoch

- Max length vega version string in upgrade proposal - enforced

- AC 1 refers to "invalid tag". What does this mean, in practice? [semmer](https://semver.org/)

- Propose and enact a downgrade - not supported - should be rejected

Need pipelines...
- Upgrade from the latest supported version, to the current version under test (more relevant for CI, once this feature is in the wild).
- Upgrade from the current version under test, to a new version (can be the same code, but which is configured to report a new protocol version. This is the best we can test for forwards compatibility ahead of actually building new versions).



- Propose with a validator which is moved to Ersatz by the time the upgrade is enacted. (What should happen here? If there are 5 validators, 3 vote yes, 2 vote no: One of the yes voters is kicked in favour of a new one, leaving the vote at 2-2. Does the upgrade still happen? - counting votes happens at block upgrade




- Ersatz validator, promoted to validator, votes to propose upgrade as deciding vote (2-2 > 3-2). Upgrade happens - counting votes happens at block upgrade

- Validator X proposes upgrade at block N, upgrade version V. Validator X _then_ proposes upgrade at block N+1, version V +/-n (some other version), other validators approve. Second upgrade succeeds. - void secondary votes

- 5 validators propose upgrade at block N. 5 propose at block N+1.  1 validator who originally proposed block N now proposes block N+1 instead. Block N upgrade is rejected. Upgrade takes place at block N+1 - event produced to show rejection

- Spam? How many transactions can be made per second from a validator? Other wallets? (raise ticket for discussion/not required at this time)


- Non-validator wallet attempts to propose upgrade. (Rejected transaction?  check tx)
- Any difference between non-validator wallet and any random vega wallet? (Not even a non-validator, just some random party). (Rejected transaction? -check tx)


- Ersatz attempts to propose upgrade. (Rejected transaction? 

- Upgrade which upgrades data node _and_ vega version. (not possible)
- Upgrade which upgrades only vega.
- Upgrade which upgrades only data node (not possible)

- Upgrade which attempts to set both tags to the current version. (not Possible)



- 3 validators propose version V at block N. 2 validators propose version V +/-n (some other version) at block N. Version V upgrade takes place. (Is it possible to propose two different versions on the same block?) - possible to vote for different tagged version across network)

No voters will stop producing blocks

- Proposal remembered in checkpoints? Proposal is scheduled for upgrade at block N. Checkpoint is taken and network is stopped at some block prior to N. Restart network from checkpoint, proposal should be remembered, upgrade takes place at block N - not applicable - votes are not carried across checkpoints

shapshot - post load - should be able to propose a vote to upgrade
Node starting from snapshot which has a proposal at a given block, ensure during replay when the block height is reached a new version is loaded and also post load an upgrade takes place at target block.


- As above, but when you restart from checkpoint, upgrade to the new version (is this valid?), such that the network is loaded with a pending proposal to the current version. What should happen? - (not applicable)


- Upgrade takes place, but insufficient validators are restored for 1, 5, 10, minutes. Validators which are restored immediately patiently wait for consensus to be achieved, and then blocks continue  - can test this.


- Upgrade takes place at block N. Minimum number of validators -1 restart with correct version. Deliberately cause one validator to restart with previous version. Consensus is not achieved. Start another validator with the correct version, consensus is achieved - no blocks produced - valid

- 5 validator network. Upgrade takes places at block N. Start 3 validators immediately. Allow several blocks to pass. Start two remaining validators. (All validators continue to work) - blocks producing if 3 have enough weight.

- 5 validator network. Upgrade takes places at block N. Start 3 validators immediately. Allow several blocks to pass. Start two remaining validators. (All validators continue to work) - no blocks producing if 3 do not have enough weight - need 70% weight to produce blocks

Test that weighting threshold (API TM) is not reached after restart, network does not produce blocks

RomanLeca commented:
One more scenario:

- there are 5 validators
- 1 validator submits an incorrect proposal (bad tag, etc)
- 3 validators submit a correct one, same version, having majority'

will this proposal pass? Yes

vega-paul commented:
Another scenario to consider to add 

Proposal upgraded requested by other validator, another validator submits multiple proposals to upgrade Y followed by N. Subsequent votes should always override previous until majority has been reached by network.

doesnt make sense - no such thing as No vote.

vega-paul committed:
Update 0075-PLUP_protocol_upgrades.md

vega-paul committed:
Update 0075-PLUP_protocol_upgrades.md

vega-paul committed:
Update 0075-PLUP_protocol_upgrades.md

vega-paul commented:
ACs updated - needs review

vega-paul committed:
Update 0075-PLUP_protocol_upgrades.md

vega-paul committed:
Update 0075-PLUP_protocol_upgrades.md

vega-paul committed:
Update 0075-PLUP_protocol_upgrades.md

jgsbennett commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,167 @@
+   -  Proposes a negative upgrade block
+   -  Proposes a 0 upgrade block
+   -  Proposes (string/other upgrade block)
+   -  Proposes with supply a block height
```
Is this supposed to be "without supplying" ?

vega-paul replied:
yes

jgsbennett commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,167 @@
+   -  Proposes an upgrade block in the past - should be rejected / failed check - sign via wallet / check wallet
+   -  Proposes a negative upgrade block
+   -  Proposes a 0 upgrade block
+   -  Proposes (string/other upgrade block)
```
@jeremyletang I think the python/protos code I've used so far tries to enforce an integer type. To what extent is it possible to send an incorrect type over the wire? i.e: By hacking together funky transactions, like with the work Muthu is doing.

jeremyletang replied:
yea well just try to build transaction yourself would be what you can do really.

jgsbennett commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,167 @@
+   
+### Epochs (<a name="0075-PLUP-006" href="#0075-PLUP-006">0075-PLUP-006</a>) 
+   - Proposing an upgrade block which ought to be the end of an epoch
+   - Propose an upgrade block which should result in a new network in the same epoch
```
These are maybe 3 different test cases, which one do we mark with the AC? Similar deal for several of the above... Our current system seems to assume one python test case per AC... For _some_ of these I could theoretically cram them into one case, but for some of the others it doesn't necessarily make sense.

vega-paul committed:
Update 0075-PLUP_protocol_upgrades.md

jgsbennett committed:
chore: Add ACs for many of the individual cases.

davidsiska-vega commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,166 @@
+   -  (<a name="0075-PLUP-006" href="#0075-PLUP-006">0075-PLUP-006</a>) Proposes a negative upgrade block
+   -  (<a name="0075-PLUP-007" href="#0075-PLUP-007">0075-PLUP-007</a>) Proposes a 0 upgrade block
+   -  (<a name="0075-PLUP-008" href="#0075-PLUP-008">0075-PLUP-008</a>) Proposes (string/other upgrade block)
+   -  (<a name="0075-PLUP-009" href="#0075-PLUP-009">0075-PLUP-009</a>) Proposes without supplying a block height
```
Should we add an AC that the block number has to be strictly greater than the current block?

gordsport replied:
@jgsbennett - please can you add this as an AC?

davidsiska-vega commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,166 @@
+   -  (<a name="0075-PLUP-009" href="#0075-PLUP-009">0075-PLUP-009</a>) Proposes without supplying a block height
+
+### VISOR 
+   - (<a name="0075-PLUP-010" href="#0075-PLUP-010">0075-PLUP-010</a>) Can be seen to automatically download the latest version for install when available at the source location when file meets the format criteria defined
```
The latest  version and the one in an accepted proposal may be different. Should we add an AC that the vegavisor downloads all the versions? Or all the ones in active proposal? Or is that not needed / intended behavior?

gordsport replied:
```suggestion
   - (<a name="0075-PLUP-010" href="#0075-PLUP-010">0075-PLUP-010</a>) Can be seen to automatically download the tagged version proposed for install when available at the source location when file meets the format criteria defined
```

davidsiska-vega commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,166 @@
+   - (<a name="0075-PLUP-011" href="#0075-PLUP-011">0075-PLUP-011</a>) Visor automatically upgrades validators to proposed version if requiredMajority has been reached
+   
+### Epochs 
+   - (<a name="0075-PLUP-012" href="#0075-PLUP-012">0075-PLUP-012</a>) Proposing an upgrade block which ought to be the end of an epoch
```
... should have what result? Is it just we want to test that if an upgrade block and epoch end coincide then all goes well?

jgsbennett replied:
We should come back up in the new epoch, and rewards should have been paid out.

gordsport replied:
Try to test this with null blockchain. If this proves to be very complex we can consider testing this later: ```suggestion
   - (<a name="0075-COSMICELEVATOR-012" href="#0075-COSMICELEVATOR-012">0075-COSMICELEVATOR-012</a>) Proposing an upgrade block which ought to be the end of an epoch
```

davidsiska-vega commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,166 @@
+The following describes the general workflow of upgrades: 
+1. Vega continuesly make new releases available 
+2. Validators can suggest to the network an upgrade with a given release tag and a block height for the upgrade to take place
+3. If a proposal gets enough votes (`protocol.upgrade.proposal.requiredMajority` of the validators, not stake) at the given height vega will take a snapshot automatically and stop processing further blocks until restarted by the vega-visor, the vega process manager. 
```
I am guessing this was discussed - but does it make sense to count validators not something like "own + delegated stake" or "tendermint weight"? Because the purpose is for the chain to continue post upgrade - so I'd think we need not only majority but the "right ones". Thoughts?

gordsport replied:
@barnabee - please can you review this and confirm if we should leave as is, or factor in the tendermint weight?

gordsport replied:
@jgsbennett - once we have a decision here can you please add an AC accordingly?

gordsport replied:
Decision in spec review meeting: We will remain as is and use count and not factor in score/weights for this milestone

jgsbennett replied:
Net param is actually validators.vote.required . Rename here and in any ACs that mention.

davidsiska-vega commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,166 @@
+### Multiple proposals (<a name="0075-PLUP-021" href="#0075-PLUP-021">0075-PLUP-021</a>)
+   - If multiple proposals are submitted from a validator before the block heights are reached then only the last proposal is considered
+
+## Spam (<a name="0075-PLUP-022" href="#0075-PLUP-022">0075-PLUP-022</a>) 
```
```suggestion
## Spam (<a name="0075-COSMICELEVATOR-022" href="#0075-COSMICELEVATOR-022">0075-COSMICELEVATOR-022</a>) 
```

davidsiska-vega commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,166 @@
+   - Excessive numbers of proposals from a single validator within an epoch should be detected and rejected - (Future requirement)
+   
+## Snapshots
+   - (<a name="0075-PLUP-023" href="#0075-PLUP-023">0075-PLUP-023</a>) Post validator joining they should be immediately allowed to propose and included in the overall total count
```
"should be immediately allowed to propose" what does propose refer to here? Propose blocks? Propose new upgrades?

jgsbennett replied:
I think the latter.

davidsiska-vega commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,166 @@
+   
+## Snapshots
+   - (<a name="0075-PLUP-023" href="#0075-PLUP-023">0075-PLUP-023</a>) Post validator joining they should be immediately allowed to propose and included in the overall total count
+   - (<a name="0075-PLUP-024" href="#0075-PLUP-024">0075-PLUP-024</a>) Validators attempting to join during upgrade will be rejected if protocol versions differ
```
Can we define "during" more precisely? Does during mean "proposal to update at block 1000 has been submitted, we are now at block 500 so no-one can join"? 

Or does "during" mean really we got to block 1000 the chain stopped while nodes restart and when it restarts it will again accept transactions and we create block 1001?

jgsbennett replied:
Again, I think the latter.

jgsbennett replied:
Drop this AC. We think it's just another specific case of the "validator joins with the wrong version" scenario.

davidsiska-vega commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,166 @@
+## Snapshots
+   - (<a name="0075-PLUP-023" href="#0075-PLUP-023">0075-PLUP-023</a>) Post validator joining they should be immediately allowed to propose and included in the overall total count
+   - (<a name="0075-PLUP-024" href="#0075-PLUP-024">0075-PLUP-024</a>) Validators attempting to join during upgrade will be rejected if protocol versions differ
+   - (<a name="0075-PLUP-025" href="#0075-PLUP-025">0075-PLUP-025</a>) Ensure that required majority is not met when enough validators join between validator proposals and target block
```
What does this mean?

jgsbennett replied:
I don't know for sure what was originally intended, but I might imagine:
1: 5 validators, 3 vote for upgrade to version V on block B.
2: Prior to that, 2 more validators join, giving total 7 validators. Block B passes without upgrade (if new validators do not vote).
(Intention is that required majority is met in step 1, but not in step 2. I haven't played with that field yet, may also require votes from the new validators to achieve this case.)

jgsbennett replied:
Clarify AC.

davidsiska-vega commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,166 @@
+   - (<a name="0075-PLUP-025" href="#0075-PLUP-025">0075-PLUP-025</a>) Ensure that required majority is not met when enough validators join between validator proposals and target block
+   - (<a name="0075-PLUP-026" href="#0075-PLUP-026">0075-PLUP-026</a>) Node starting from snapshot which has a proposal at a given block, ensure during replay when the block height is reached a new version is loaded and also post load an upgrade takes place at target block.
+
+## Checkpoints
```
```suggestion
## LNL Checkpoints
```

davidsiska-vega commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,166 @@
+ 
+ 
+ ### Failing consensus (<a name="0075-PLUP-031" href="#0075-PLUP-031">0075-PLUP-031</a>)
+  - Upgrade takes place at block N. Minimum number of validators -1 restart with correct version. One validator restarts with previous version. Consensus is not achieved. Start another validator with the correct version, consensus is achieved
```
Regardless of what own stake + delegations they have?

jgsbennett replied:
Step 1:
Restart the network with a number of validators whose voting power is equal to less than two thirds. Consensus is not achieved.
Step 2:
Restart one of the validators (whose voting power would take the total voting power over the two thirds threshold) with the wrong version. Consensus should still not be achieved.
Step 3:
Restart validator from step 2 with the correct version. Consensus is achieved.

davidsiska-vega reviewed (changes requested):
Left a lot of comments - some of them may stem from my ignorance of the design - if that's the case say so and mark as resolved ;-).

gordsport committed:
chore: Update protocol/0075-PLUP_protocol_upgrades.md

Co-authored-by: David Siska <62546419+davidsiska-vega@users.noreply.github.com>

gordsport commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,166 @@
+   - Excessive numbers of proposals from a single validator within an epoch should be detected and rejected - (Future requirement)
+   
+## Snapshots
+   - (<a name="0075-PLUP-023" href="#0075-PLUP-023">0075-PLUP-023</a>) Post validator joining they should be immediately allowed to propose and included in the overall total count
```
```suggestion
   - (<a name="0075-PLUP-023" href="#0075-PLUP-023">0075-PLUP-023</a>) Post a validator becoming a tendermint validator they should be immediately allowed to propose an upgrade and be included in the overall total count
```

jgsbennett committed:
chore: Update protocol/0075-PLUP_protocol_upgrades.md

Co-authored-by: David Siska <62546419+davidsiska-vega@users.noreply.github.com>

jgsbennett commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,166 @@
+ ### Failing consensus (<a name="0075-PLUP-031" href="#0075-PLUP-031">0075-PLUP-031</a>)
+  - Upgrade takes place at block N. Minimum number of validators -1 restart with correct version. One validator restarts with previous version. Consensus is not achieved. Start another validator with the correct version, consensus is achieved
+  - 5 validator network. Upgrade takes places at block N. Start 3 validators immediately. Allow several blocks to pass. Start two remaining validators. (All validators continue to work) - no blocks producing as 3 validators do not have enough weight - need 70% weight to produce blocks
+  - Weighting threshold (API TM) is not reached after restart, network does not produce blocks
```
Drop

jgsbennett commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,166 @@
+  - Upgrade takes place at block N. Minimum number of validators -1 restart with correct version. One validator restarts with previous version. Consensus is not achieved. Start another validator with the correct version, consensus is achieved
+  - 5 validator network. Upgrade takes places at block N. Start 3 validators immediately. Allow several blocks to pass. Start two remaining validators. (All validators continue to work) - no blocks producing as 3 validators do not have enough weight - need 70% weight to produce blocks
+  - Weighting threshold (API TM) is not reached after restart, network does not produce blocks
+  - Upgrade takes place, but insufficient validators are restored for 1, 5, 10, minutes. Validators which are restored immediately patiently wait for consensus to be achieved, and then blocks continue  - concensus acheived
```
Add an AC for the mainnet test case. https://github.com/vegaprotocol/devops-infra/issues/1522

jgsbennett replied:
Add an AC for "submit a proposal with version == current version and a valid block height retracts any existing proposals"
Add an AC for "submit an invalid proposal does not overwrite any existing proposals" i.e: Set up an upgrade set to happen at block height H. One validator (whose changing vote would cause the upgrade to not happen) submits an invalid upgrade proposal (i.e: invalid version). Proposal rejected. Upgrade still happens.

jgsbennett committed:
Update protocol/0075-PLUP_protocol_upgrades.md

Co-authored-by: Gordsport <83510148+gordsport@users.noreply.github.com>

jgsbennett committed:
Update protocol/0075-PLUP_protocol_upgrades.md

Co-authored-by: Gordsport <83510148+gordsport@users.noreply.github.com>

jgsbennett committed:
fix: Review comments, clarify ACs and add some more.

jgsbennett committed:
fix: Update network param used to judge required majority.

jgsbennett committed:
fix: Add two more cases for the required majority network parameter.

davidsiska-vega commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,174 @@
+   
+### Epochs 
+   - (<a name="(0075-COSMICELEVATOR-012)" href="#(0075-COSMICELEVATOR-012)">(0075-COSMICELEVATOR-012)</a>) Proposing an upgrade block which ought to be the end of an epoch. After upgrade takes place, confirm rewards are distributed, any pending delegations take effect, and validator joining/leaving takes effect.
+   - (<a name="0075-PLUP-013" href="#0075-PLUP-013">0075-PLUP-013</a>) Propose an upgrade block which should result in a new network in the same epoch
```
```suggestion
   - (<a name="0075-PLUP-013" href="#0075-PLUP-013">0075-PLUP-013</a>) Propose an upgrade block which should result in a network running new code version in the same epoch.
```

davidsiska-vega commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,174 @@
+   - (<a name="0075-PLUP-013" href="#0075-PLUP-013">0075-PLUP-013</a>) Propose an upgrade block which should result in a new network in the same epoch
+   - (<a name="0075-PLUP-014" href="#0075-PLUP-014">0075-PLUP-014</a>) Ensure end of epoch processes still run after restore e.g reward calculation and distributions
+
+### Required Majority
```
```suggestion
### Required Majority
For the purposes of protocol upgrade each validator that participates in consensus has one vote. Required majority is set by `validators.vote.required network parameter`. 
```

davidsiska-vega commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,174 @@
+
+### Required Majority
+   - (<a name="0075-PLUP-015" href="#0075-PLUP-015">0075-PLUP-015</a>) Counting proposal votes to check if required majority has been reached occurs when any proposed target block has been reached
+   - (<a name="0075-PLUP-016" href="#0075-PLUP-016">0075-PLUP-016</a>) Only active network validators proposals are counted when any proposed target block has been reached 
```
```suggestion
   - (<a name="0075-PLUP-016" href="#0075-PLUP-016">0075-PLUP-016</a>) Only proposals from validators participating in consensus are counted when any proposed target block has been reached. 
```

davidsiska-vega commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,174 @@
+   - (<a name="0075-PLUP-018" href="#0075-PLUP-018">0075-PLUP-018</a>) When majority reached during the process of upgrading, those validators which didnt propose will stop producing blocks
+   - (<a name="0075-PLUP-019" href="#0075-PLUP-019">0075-PLUP-019</a>) Proposals for multiple versions at same block height will be rejected if majority has not been reached, network continues with the current running version 
+   - (<a name="0075-PLUP-020" href="#0075-PLUP-020">0075-PLUP-020</a>) Propose with a validator which is moved to Ersatz by the time the upgrade is enacted. If there are 5 validators, 3 vote yes, 2 vote no: One of the yes voters is kicked in favour of a new one, leaving the vote at 2-2 so the upgrade should not happen as counting votes happens at block height only
+   - (<a name="0075-PLUP-036" href="#0075-PLUP-036">0075-PLUP-036</a>) Changing validators.vote.required network parameter to a value above two thirds is respected.
```
```suggestion
   - (<a name="0075-PLUP-036" href="#0075-PLUP-036">0075-PLUP-036</a>) Changing `validators.vote.required` network parameter to a value above two thirds is respected.
```

davidsiska-vega commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,174 @@
+   - (<a name="0075-PLUP-019" href="#0075-PLUP-019">0075-PLUP-019</a>) Proposals for multiple versions at same block height will be rejected if majority has not been reached, network continues with the current running version 
+   - (<a name="0075-PLUP-020" href="#0075-PLUP-020">0075-PLUP-020</a>) Propose with a validator which is moved to Ersatz by the time the upgrade is enacted. If there are 5 validators, 3 vote yes, 2 vote no: One of the yes voters is kicked in favour of a new one, leaving the vote at 2-2 so the upgrade should not happen as counting votes happens at block height only
+   - (<a name="0075-PLUP-036" href="#0075-PLUP-036">0075-PLUP-036</a>) Changing validators.vote.required network parameter to a value above two thirds is respected.
+   - (<a name="0075-PLUP-037" href="#0075-PLUP-037">0075-PLUP-037</a>) The value of validators.vote.required is checked at upgrade block, i.e: vote on a proposal with all validators, then change the validators.vote.required net param before upgrade block, to a higher value, which would cause the upgrade to be rejected. Upgrade fails.
```
```suggestion
   - (<a name="0075-PLUP-037" href="#0075-PLUP-037">0075-PLUP-037</a>) The value of `validators.vote.required` is checked at upgrade block, i.e: vote on a proposal with all validators, then change the `validators.vote.required` net param before upgrade block, to a higher value, which would cause the upgrade to be rejected. Upgrade fails.
```

davidsiska-vega commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,174 @@
+   - Excessive numbers of proposals from a single validator within an epoch should be detected and rejected - (Future requirement)
+   
+## Snapshots
+   - (<a name="0075-PLUP-023" href="#0075-PLUP-023">0075-PLUP-023</a>) Post a validator becoming a tendermint validator they should be immediately allowed to propose an upgrade and be included in the overall total count
```
```suggestion
   - (<a name="0075-PLUP-023" href="#0075-PLUP-023">0075-PLUP-023</a>) Post a validator becoming a consensus-participating validator they should be immediately allowed to propose an upgrade and be included in the overall total count
```

davidsiska-vega commented on protocol/0075-PLUP_protocol_upgrades.md:
```diff
@@ -0,0 +1,174 @@
+ 
+### Failing consensus
+  - (<a name="0075-PLUP-030" href="#0075-PLUP-030">0075-PLUP-030</a>) Upgrade takes place at block N. Restart with a number of validators whose voting power is <= two thirds. Restart one more validator whose voting power would take the total voting power >= two thirds, with an incorrect version. Consensus is not achieved. Now restart that validator with the correct version. Consensus is achieved.
+  - (<a name="0075-PLUP-031" href="#0075-PLUP-031">0075-PLUP-031</a>) 5 validator network. Upgrade takes places at block N. Start 3 validators immediately. Allow several blocks to pass. Start two remaining validators. (All validators continue to work) - no blocks producing as 3 validators do not have enough weight - need 70% weight to produce blocks
```
Allow several blocks to pass. ... no blocks producing as 3 validators do not have enough weight  

This is confusing. Do we mean allow *time* to pass first.

jgsbennett replied:
Yes, I never saw that. I'll correct it.

davidsiska-vega committed:
chore: Apply suggestions from code review

davidsiska-vega reviewed (changes requested):
One clarification - then done.

jgsbennett committed:
fix: AC incorrectly implied blocks can be created without consensus.
